File: leetcode/lc-us/cpp/6.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    string convert(string s, int numRows) {
        if (s.length() <= numRows || numRows == 1) return s;
        vector<string> rows(numRows);
        bool downward = false;
        int cur = 0;
        for (char c: s) {
            if (cur == 0 || cur == numRows - 1) downward = !downward;
            rows[cur].push_back(c);
            cur += downward ? 1 : -1;
        }
        string res;
        for (const string &row: rows) res.append(row);
        return res;
    }
};

File: leetcode/lc-us/py/6.py
class Solution:
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1:
            return s
        v = ['' for _ in range(numRows)]
        row, dt = 0, 1
        for i in range(len(s)):
            v[row] += s[i]
            row += dt
            if row == 0 or row == numRows - 1:
                dt = -dt
        return ''.join(v)
